core/pta/attestation.c

```diff
@@ -0,0 +1,600 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020, Elias von Däniken
+ */
+
+#include <compiler.h>
+#include <stdio.h>
+#include <kernel/pseudo_ta.h>
+#include <mm/tee_pager.h>
+#include <mm/tee_mm.h>
+#include <pta_attestation.h>
+#include <string.h>
+#include <string_ext.h>
+#include <malloc.h>
+#include <tee_api_types.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+
+#define RSA_SIG_SIZE_BYTE 256
+
+/*
+ * Attestation data is static to keep until the device boots again.
+ * Afterwards the attestation data has to be reloaded and everthing
+ * has to be attestated newly.
+ */
+static struct attestation_data {
+	uint8_t *iv;
+	uint32_t iv_size;
+	uint8_t *pub;
+	uint32_t pub_size;
+	uint8_t *enc_priv;
+	uint32_t enc_priv_size;
+	uint8_t *sig;
+	uint32_t sig_size;
+} att_data;
+
+static TEE_Result priv_asn1_decode(uint8_t *priv,
+				uint8_t **d, size_t *d_size)
+{
+
+	uint32_t size;
+	uint32_t offset = 7; // Beginn at the start of the modulus
+
+	// Check the header
+	if (priv[0] != 0x30)
+		return TEE_ERROR_SECURITY;
+
+	if (priv[1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	// Get the size of the first integer aka modulus
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	if (priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (priv[offset+2] << 8) | priv[offset+3];
+	offset += size + 4;
+
+	// Get the size of the second integer aka exponent
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	size = priv[offset + 1];
+	offset += size + 2;
+
+	// Decoding d
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	if (priv[offset+1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (priv[offset+2] << 8) | priv[offset+3];
+	*d = &(priv[offset+4]);
+	*d_size = size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pub_asn1_decode(uint8_t **n, uint8_t **e,
+				size_t *n_size, size_t *e_size)
+{
+	uint32_t size;
+
+	// Decoding header
+	if (att_data.pub[0] != 0x30)
+		return TEE_ERROR_SECURITY;
+
+	if (att_data.pub[1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[2] << 8) | att_data.pub[3];
+	if ((size + 4) != att_data.pub_size)
+		return TEE_ERROR_SECURITY;
+
+
+	// Decoding first integer aka modulus
+	if (att_data.pub[4] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	if (att_data.pub[5] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[6] << 8) | att_data.pub[7];
+	*n = &(att_data.pub[8]);
+	*n_size = size;
+
+	// Decoding second integer aka exponent
+	if (att_data.pub[size + 8] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	*e_size = att_data.pub[size + 9];
+	*e = &(att_data.pub[size + 10]);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_sys_measurement(uint8_t *ptr)
```

Comment:
This should probably be global and `__weak` in order to be easy to override in platform code.

Reply:
Good point, but I am not sure where to put the global definitions. The header file in the libutee is obviously the wrong place.

Should the header file be next to attestation.c ?